File: ts/src/trace.ts
/**
 * Per-stage observability — parity port of contexel's Python `trace`.
 * `trace({idField}, fn)` runs `fn` with an active trace (AsyncLocalStorage,
 * the JS analog of contextvars) and records what every stage did; with
 * `idField` it also records which records each stage dropped (records
 * lacking the field are untracked). `audit()` is the governance record.
 */
import { AsyncLocalStorage } from "node:async_hooks";
import * as tokens from "./tokens.js";
import { pyRound } from "./py.js";

export interface Entry {
  stage: string;
  inCount: number;
  outCount: number;
  tokensBefore: number;
  tokensAfter: number;
  droppedIds: unknown[];
}

export class Trace {
  entries: Entry[] = [];
  policies: string[] = [];
  idField: string | null;

  constructor(idField: string | null = null) {
    this.idField = idField;
  }

  get tokensBefore(): number {
    return this.entries.length ? this.entries[0].tokensBefore : 0;
  }

  get tokensAfter(): number {
    return this.entries.length ? this.entries[this.entries.length - 1].tokensAfter : 0;
  }

  get reduction(): number {
    const before = this.tokensBefore;
    return before === 0 ? 0 : 1 - this.tokensAfter / before;
  }

  report(): string {
    if (!this.entries.length) return "(no stages traced)";
    const lines = this.entries.map(
      (e) =>
        `${e.stage.padEnd(16)} ${String(e.inCount).padStart(5)} -> ` +
        `${String(e.outCount).padEnd(5)} ${e.tokensBefore.toLocaleString("en-US").padStart(7)} -> ` +
        `${e.tokensAfter.toLocaleString("en-US").padEnd(7)} tokens`
    );
    lines.push("-".repeat(56));
    lines.push(
      `${"total".padEnd(16)} ${"".padStart(5)}    ${"".padEnd(5)} ` +
        `${this.tokensBefore.toLocaleString("en-US").padStart(7)} -> ` +
        `${this.tokensAfter.toLocaleString("en-US").padEnd(7)} tokens  ` +
        `(${pyRound(this.reduction * 100, 0)}% reduction)`
    );
    return lines.join("\n");
  }

  /** JSON-able governance record: policy fingerprints, per-stage dropped
   * record ids (stage = removal reason; records lacking the field are
   * untracked), token movement. */
  audit(): Record<string, unknown> {
    return {
      policies: [...this.policies],
      id_field: this.idField,
      tokens_before: this.tokensBefore,
      tokens_after: this.tokensAfter,
      reduction: pyRound(this.reduction, 4), // Python round(): half to even
      stages: this.entries.map((e) => ({
        stage: e.stage,
        records_in: e.inCount,
        records_out: e.outCount,
        tokens_before: e.tokensBefore,
        tokens_after: e.tokensAfter,
        dropped_ids: [...e.droppedIds],
      })),
    };
  }
}

const store = new AsyncLocalStorage<Trace>();

export function currentTrace(): Trace | null {
  return store.getStore() ?? null;
}

/** Run `fn` with an active trace; returns [result, trace]. */
export function trace<T>(
  options: { idField?: string } | null,
  fn: () => T
): [T, Trace] {
  const tr = new Trace(options?.idField ?? null);
  const result = store.run(tr, fn);
  return [result, tr];
}

type Rec = Record<string, unknown>;

function idKey(value: unknown): unknown {
  if (value !== null && typeof value === "object") return tokens.pythonJson(value);
  return value;
}

/** Wrap a stage so an active trace records its effect (incl. dropped ids). */
export function traced<A extends unknown[]>(
  name: string,
  fn: (records: Rec[], ...args: A) => Rec[]
): (records: Rec[], ...args: A) => Rec[] {
  return (records: Rec[], ...args: A): Rec[] => {
    const tr = currentTrace();
    if (!tr) return fn(records, ...args);
    const beforeN = records.length;
    const beforeTok = tokens.count(records);
    const out = fn(records, ...args);
    const dropped: unknown[] = [];
    const fid = tr.idField;
    if (fid) {
      const remaining = new Map<unknown, number>();
      let unidentified = 0; // survivors without the id field absorb drops
      for (const r of out) {
        if (r !== null && typeof r === "object" && Object.hasOwn(r, fid)) {
          const k = idKey((r as Rec)[fid]);
          remaining.set(k, (remaining.get(k) ?? 0) + 1);
        } else {
          unidentified++;
        }
      }
      for (const r of records) {
        if (r !== null && typeof r === "object" && Object.hasOwn(r, fid)) {
          const original = (r as Rec)[fid];
          const k = idKey(original);
          const left = remaining.get(k) ?? 0;
          if (left > 0) remaining.set(k, left - 1);
          else if (unidentified > 0) unidentified--;
          else dropped.push(original);
        }
      }
    }
    tr.entries.push({
      stage: name,
      inCount: beforeN,
      outCount: out.length,
      tokensBefore: beforeTok,
      tokensAfter: tokens.count(out),
      droppedIds: dropped,
    });
    return out;
  };
}

File: csharp/Contexel/Trace.cs
using System.Globalization;
using System.Text;

namespace Contexel;

using Rec = Dictionary<string, object?>;
using Records = List<Dictionary<string, object?>>;

/// <summary>
/// Per-stage observability — parity port of contexel's Python trace.
/// <c>using var t = Trace.Begin(idField: "path")</c> activates a trace
/// (AsyncLocal, the .NET analog of contextvars); stages auto-record their
/// effect, including which records each stage dropped. <c>Audit()</c> is
/// the governance record (snake_case keys, matching the golden vectors).
/// </summary>
public sealed class Trace : IDisposable
{
    public sealed record Entry(string Stage, int InCount, int OutCount,
                               int TokensBefore, int TokensAfter,
                               List<object?> DroppedIds);

    public List<Entry> Entries { get; } = [];
    public List<string> Policies { get; } = [];
    public string? IdField { get; }

    private static readonly AsyncLocal<Trace?> CurrentLocal = new();
    private readonly Trace? _previous;

    private Trace(string? idField)
    {
        IdField = idField;
        _previous = CurrentLocal.Value;
        CurrentLocal.Value = this;
    }

    public static Trace Begin(string? idField = null) => new(idField);
    public static Trace? Current => CurrentLocal.Value;

    public void Dispose() => CurrentLocal.Value = _previous;

    public int TokensBefore => Entries.Count > 0 ? Entries[0].TokensBefore : 0;
    public int TokensAfter => Entries.Count > 0 ? Entries[^1].TokensAfter : 0;
    public double Reduction =>
        TokensBefore == 0 ? 0 : 1 - (double)TokensAfter / TokensBefore;

    public string Report()
    {
        if (Entries.Count == 0) return "(no stages traced)";
        var lines = Entries.Select(e =>
            $"{e.Stage,-16} {e.InCount,5} -> {e.OutCount,-5} " +
            $"{e.TokensBefore.ToString("N0", CultureInfo.InvariantCulture),7} -> " +
            $"{e.TokensAfter.ToString("N0", CultureInfo.InvariantCulture),-7} tokens").ToList();
        lines.Add(new string('-', 56));
        lines.Add(
            $"{"total",-16} {"",5}    {"",-5} " +
            $"{TokensBefore.ToString("N0", CultureInfo.InvariantCulture),7} -> " +
            $"{TokensAfter.ToString("N0", CultureInfo.InvariantCulture),-7} tokens  " +
            $"({PyCompat.PyRound(Reduction * 100, 0).ToString(CultureInfo.InvariantCulture)}% reduction)");
        return string.Join("\n", lines);
    }

    /// <summary>JSON-able governance record: policy fingerprints,
    /// per-stage dropped ids (stage = removal reason), token movement.</summary>
    public Rec Audit() => new()
    {
        ["policies"] = new List<object?>(Policies),
        ["id_field"] = IdField,
        ["tokens_before"] = (long)TokensBefore,
        ["tokens_after"] = (long)TokensAfter,
        ["reduction"] = PyCompat.PyRound(Reduction, 4),
        ["stages"] = Entries.Select(e => (object?)new Rec
        {
            ["stage"] = e.Stage,
            ["records_in"] = (long)e.InCount,
            ["records_out"] = (long)e.OutCount,
            ["tokens_before"] = (long)e.TokensBefore,
            ["tokens_after"] = (long)e.TokensAfter,
            ["dropped_ids"] = new List<object?>(e.DroppedIds),
        }).ToList(),
    };

    /// <summary>Python-dict-faithful id key: 1, 1.0, and True share a key
    /// (Python dict equality), containers key by canonical JSON.</summary>
    private static string IdKey(object? value) => value switch
    {
        null => "None",
        bool or long or int or double => PyCompat.PyEqKey(value),
        string s => "str:" + s,
        _ => "json:" + Tokens.PythonJson(value),
    };

    /// <summary>Wrap a stage body so an active trace records its effect
    /// (including dropped ids, with the unidentified-survivor wildcard).</summary>
    internal static Records Traced(string name, Records records, Func<Records> fn)
    {
        var tr = Current;
        if (tr is null) return fn();
        int beforeN = records.Count;
        int beforeTok = Tokens.Count(records);
        var output = fn();
        var dropped = new List<object?>();
        string? fid = tr.IdField;
        if (!string.IsNullOrEmpty(fid)) // Python: `if self.id_field:` — "" is falsy
        {
            var remaining = new Dictionary<string, int>();
            int unidentified = 0; // survivors without the id field absorb drops
            foreach (var r in output)
            {
                if (r.TryGetValue(fid, out var v))
                {
                    string k = IdKey(v);
                    remaining[k] = remaining.GetValueOrDefault(k) + 1;
                }
                else unidentified++;
            }
            foreach (var r in records)
            {
                if (r.TryGetValue(fid, out var original))
                {
                    string k = IdKey(original);
                    int left = remaining.GetValueOrDefault(k);
                    if (left > 0) remaining[k] = left - 1;
                    else if (unidentified > 0) unidentified--;
                    else dropped.Add(original);
                }
            }
        }
        tr.Entries.Add(new Entry(name, beforeN, output.Count, beforeTok,
                                 Tokens.Count(output), dropped));
        return output;
    }
}
